File: backend/routers/agents.py
"""Agent CRUD API endpoints."""
import logging
from pathlib import Path
from fastapi import APIRouter
from pydantic import BaseModel
from schemas.agent import AgentCreateRequest, AgentUpdateRequest, AgentResponse
from database import db
from config import ANTHROPIC_TO_BEDROCK_MODEL_MAP
from core.exceptions import (
    AgentNotFoundException,
    ValidationException,
)
from core.workspace_manager import workspace_manager


class WorkingDirectoryResponse(BaseModel):
    """Response model for agent working directory."""
    path: str
    is_global_mode: bool

logger = logging.getLogger(__name__)

router = APIRouter()


@router.get("/models", response_model=list[str])
async def list_available_models():
    """List all available Claude models.

    Returns the Anthropic model IDs that have Bedrock mappings configured.
    """
    return list(ANTHROPIC_TO_BEDROCK_MODEL_MAP.keys())


@router.get("", response_model=list[AgentResponse])
async def list_agents():
    """List all agents."""
    agents = await db.agents.list()
    # Filter out the default agent from the list
    return [a for a in agents if a.get("id") != "default"]


@router.get("/default", response_model=AgentResponse)
async def get_default_agent():
    """Get the default system agent."""
    agent = await db.agents.get("default")
    if not agent:
        raise AgentNotFoundException(
            detail="Default agent configuration is missing",
            suggested_action="Contact the administrator to set up the default agent"
        )
    return agent


@router.get("/{agent_id}", response_model=AgentResponse)
async def get_agent(agent_id: str):
    """Get a specific agent by ID."""
    agent = await db.agents.get(agent_id)
    if not agent:
        raise AgentNotFoundException(
            detail=f"Agent with ID '{agent_id}' does not exist",
            suggested_action="Please check the agent ID and try again"
        )
    return agent


@router.get("/{agent_id}/working-directory", response_model=WorkingDirectoryResponse)
async def get_agent_working_directory(agent_id: str):
    """Get the effective working directory for an agent.

    Returns the agent's working directory based on its configuration:
    - Global User Mode: Returns home directory (~/)
    - Isolated Mode: Returns the per-agent workspace directory

    Note: If a session-level workDir is set (from "work in a folder"),
    that should override this on the frontend.
    """
    agent = await db.agents.get(agent_id)
    if not agent:
        raise AgentNotFoundException(
            detail=f"Agent with ID '{agent_id}' does not exist",
            suggested_action="Please check the agent ID and try again"
        )

    global_user_mode = agent.get("global_user_mode", True)  # Default to True now

    if global_user_mode:
        # Global User Mode: use home directory
        working_dir = str(Path.home())
    else:
        # Isolated Mode: use per-agent workspace
        working_dir = str(workspace_manager.get_agent_workspace(agent_id))

    return WorkingDirectoryResponse(
        path=working_dir,
        is_global_mode=global_user_mode
    )


@router.post("", response_model=AgentResponse, status_code=201)
async def create_agent(request: AgentCreateRequest):
    """Create a new agent."""
    # Global User Mode requires allow_all_skills=True (skill restrictions not supported)
    global_user_mode = request.global_user_mode
    allow_all_skills = request.allow_all_skills
    skill_ids = request.skill_ids

    if global_user_mode:
        allow_all_skills = True
        skill_ids = []  # Clear skill_ids since all skills are allowed
        logger.info("Global User Mode enabled - setting allow_all_skills=True, clearing skill_ids")

    agent_data = {
        "name": request.name,
        "description": request.description,
        "model": request.model,
        "permission_mode": request.permission_mode,
        "max_turns": request.max_turns,
        "system_prompt": request.system_prompt,
        "allowed_tools": request.allowed_tools,
        "plugin_ids": request.plugin_ids,
        "skill_ids": skill_ids,
        "allow_all_skills": allow_all_skills,
        "mcp_ids": request.mcp_ids,
        "working_directory": None,  # Use default from settings.agent_workspace_dir
        "enable_bash_tool": request.enable_bash_tool,
        "enable_file_tools": request.enable_file_tools,
        "enable_web_tools": request.enable_web_tools,
        "enable_tool_logging": True,
        "enable_safety_checks": True,
        "enable_file_access_control": request.enable_file_access_control,
        "allowed_directories": request.allowed_directories,
        "global_user_mode": global_user_mode,
        "enable_human_approval": request.enable_human_approval,
        "status": "active",
    }
    agent = await db.agents.put(agent_data)

    # Build per-agent workspace with symlinks to allowed skills
    try:
        await workspace_manager.rebuild_agent_workspace(
            agent_id=agent["id"],
            skill_ids=request.skill_ids,
            allow_all_skills=request.allow_all_skills
        )
        logger.info(f"Created workspace for agent {agent['id']}")
    except Exception as e:
        logger.error(f"Failed to create workspace for agent {agent['id']}: {e}")
        # Don't fail agent creation if workspace creation fails

    return agent


@router.put("/{agent_id}", response_model=AgentResponse)
async def update_agent(agent_id: str, request: AgentUpdateRequest):
    """Update an existing agent."""
    existing = await db.agents.get(agent_id)
    if not existing:
        raise AgentNotFoundException(
            detail=f"Agent with ID '{agent_id}' does not exist",
            suggested_action="Please check the agent ID and try again"
        )

    updates = request.model_dump(exclude_unset=True)

    # Global User Mode requires allow_all_skills=True (skill restrictions not supported)
    # Check if global_user_mode is being set or was already set
    global_user_mode = updates.get("global_user_mode", existing.get("global_user_mode", False))

    if global_user_mode:
        updates["allow_all_skills"] = True
        updates["skill_ids"] = []  # Clear skill_ids since all skills are allowed
        logger.info(f"Global User Mode enabled for agent {agent_id} - setting allow_all_skills=True, clearing skill_ids")

    agent = await db.agents.update(agent_id, updates)

    # Check if skill_ids or allow_all_skills changed - if so, rebuild workspace
    skill_ids_changed = "skill_ids" in updates
    allow_all_changed = "allow_all_skills" in updates

    if skill_ids_changed or allow_all_changed:
        try:
            skill_ids = agent.get("skill_ids", [])
            allow_all_skills = agent.get("allow_all_skills", False)
            await workspace_manager.rebuild_agent_workspace(
                agent_id=agent_id,
                skill_ids=skill_ids,
                allow_all_skills=allow_all_skills
            )
            logger.info(f"Rebuilt workspace for agent {agent_id} after skill config change")
        except Exception as e:
            logger.error(f"Failed to rebuild workspace for agent {agent_id}: {e}")
            # Don't fail agent update if workspace rebuild fails

    return agent


@router.delete("/{agent_id}", status_code=204)
async def delete_agent(agent_id: str):
    """Delete an agent."""
    if agent_id == "default":
        raise ValidationException(
            message="Cannot delete the default agent",
            detail="The default agent is a system resource and cannot be deleted",
            suggested_action="If you need to modify the default agent, use the update endpoint instead"
        )

    deleted = await db.agents.delete(agent_id)
    if not deleted:
        raise AgentNotFoundException(
            detail=f"Agent with ID '{agent_id}' does not exist",
            suggested_action="Please check the agent ID and try again"
        )

    # Clean up agent workspace
    try:
        await workspace_manager.delete_agent_workspace(agent_id)
        logger.info(f"Deleted workspace for agent {agent_id}")
    except Exception as e:
        logger.error(f"Failed to delete workspace for agent {agent_id}: {e}")
        # Don't fail agent deletion if workspace cleanup fails

File: desktop/src/services/agents.ts
import api from './api';
import type { Agent, AgentCreateRequest, AgentUpdateRequest, SandboxConfig, SandboxConfigRequest } from '../types';

// Convert sandbox network config to snake_case
const sandboxNetworkToSnakeCase = (network: SandboxConfigRequest['network']): Record<string, unknown> | undefined => {
  if (!network) return undefined;
  const result: Record<string, unknown> = {};
  if (network.allowLocalBinding !== undefined) result.allow_local_binding = network.allowLocalBinding;
  if (network.allowUnixSockets !== undefined) result.allow_unix_sockets = network.allowUnixSockets;
  if (network.allowAllUnixSockets !== undefined) result.allow_all_unix_sockets = network.allowAllUnixSockets;
  return Object.keys(result).length > 0 ? result : undefined;
};

// Convert sandbox config to snake_case
const sandboxToSnakeCase = (sandbox: SandboxConfigRequest): Record<string, unknown> => {
  const result: Record<string, unknown> = {};
  if (sandbox.enabled !== undefined) result.enabled = sandbox.enabled;
  if (sandbox.autoAllowBashIfSandboxed !== undefined) result.auto_allow_bash_if_sandboxed = sandbox.autoAllowBashIfSandboxed;
  if (sandbox.excludedCommands !== undefined) result.excluded_commands = sandbox.excludedCommands;
  if (sandbox.allowUnsandboxedCommands !== undefined) result.allow_unsandboxed_commands = sandbox.allowUnsandboxedCommands;
  const networkResult = sandboxNetworkToSnakeCase(sandbox.network);
  if (networkResult) result.network = networkResult;
  return result;
};

// Convert sandbox config from snake_case to camelCase
const sandboxToCamelCase = (data: Record<string, unknown>): SandboxConfig => {
  const networkData = data.network as Record<string, unknown> | undefined;
  return {
    enabled: (data.enabled as boolean) ?? false,
    autoAllowBashIfSandboxed: (data.auto_allow_bash_if_sandboxed as boolean) ?? true,
    excludedCommands: (data.excluded_commands as string[]) ?? [],
    allowUnsandboxedCommands: (data.allow_unsandboxed_commands as boolean) ?? false,
    network: {
      allowLocalBinding: (networkData?.allow_local_binding as boolean) ?? false,
      allowUnixSockets: (networkData?.allow_unix_sockets as string[]) ?? [],
      allowAllUnixSockets: (networkData?.allow_all_unix_sockets as boolean) ?? false,
    },
  };
};

// Convert camelCase to snake_case for API requests
const toSnakeCase = (data: AgentCreateRequest | AgentUpdateRequest) => {
  const result: Record<string, unknown> = {};
  if (data.name !== undefined) result.name = data.name;
  if (data.description !== undefined) result.description = data.description;
  if (data.model !== undefined) result.model = data.model;
  if (data.permissionMode !== undefined) result.permission_mode = data.permissionMode;
  if (data.systemPrompt !== undefined) result.system_prompt = data.systemPrompt;
  if (data.pluginIds !== undefined) result.plugin_ids = data.pluginIds;
  if (data.skillIds !== undefined) result.skill_ids = data.skillIds;
  if (data.allowAllSkills !== undefined) result.allow_all_skills = data.allowAllSkills;
  if (data.mcpIds !== undefined) result.mcp_ids = data.mcpIds;
  if (data.allowedTools !== undefined) result.allowed_tools = data.allowedTools;
  if (data.enableBashTool !== undefined) result.enable_bash_tool = data.enableBashTool;
  if (data.enableFileTools !== undefined) result.enable_file_tools = data.enableFileTools;
  if (data.enableWebTools !== undefined) result.enable_web_tools = data.enableWebTools;
  if (data.globalUserMode !== undefined) result.global_user_mode = data.globalUserMode;
  if (data.enableHumanApproval !== undefined) result.enable_human_approval = data.enableHumanApproval;
  if (data.sandbox !== undefined) result.sandbox = sandboxToSnakeCase(data.sandbox);
  return result;
};

// Convert snake_case response to camelCase
const toCamelCase = (data: Record<string, unknown>): Agent => {
  // Parse sandbox config (optional - may be undefined for legacy agents)
  const sandboxData = data.sandbox as Record<string, unknown> | undefined;
  const sandbox: SandboxConfig | undefined = sandboxData
    ? sandboxToCamelCase(sandboxData)
    : undefined;

  return {
    id: data.id as string,
    name: data.name as string,
    description: data.description as string | undefined,
    model: data.model as string | undefined,
    permissionMode: (data.permission_mode as Agent['permissionMode']) ?? 'default',
    systemPrompt: data.system_prompt as string | undefined,
    allowedTools: (data.allowed_tools as string[]) || [],
    pluginIds: (data.plugin_ids as string[]) || [],
    skillIds: (data.skill_ids as string[]) || [],
    allowAllSkills: (data.allow_all_skills as boolean) ?? false,
    mcpIds: (data.mcp_ids as string[]) || [],
    workingDirectory: data.working_directory as string | undefined,
    enableBashTool: (data.enable_bash_tool as boolean) ?? true,
    enableFileTools: (data.enable_file_tools as boolean) ?? true,
    enableWebTools: (data.enable_web_tools as boolean) ?? false,
    enableToolLogging: (data.enable_tool_logging as boolean) ?? true,
    enableSafetyChecks: (data.enable_safety_checks as boolean) ?? true,
    globalUserMode: (data.global_user_mode as boolean) ?? false,
    enableHumanApproval: (data.enable_human_approval as boolean) ?? true,
    sandbox,
    status: (data.status as 'active' | 'inactive') ?? 'active',
    createdAt: (data.created_at as string) ?? '',
    updatedAt: (data.updated_at as string) ?? '',
  };
};

export const agentsService = {
  // List all agents
  async list(): Promise<Agent[]> {
    const response = await api.get<Record<string, unknown>[]>('/agents');
    return response.data.map(toCamelCase);
  },

  // Get agent by ID
  async get(id: string): Promise<Agent> {
    const response = await api.get<Record<string, unknown>>(`/agents/${id}`);
    return toCamelCase(response.data);
  },

  // Create new agent
  async create(data: AgentCreateRequest): Promise<Agent> {
    const response = await api.post<Record<string, unknown>>('/agents', toSnakeCase(data));
    return toCamelCase(response.data);
  },

  // Update agent
  async update(id: string, data: AgentUpdateRequest): Promise<Agent> {
    const response = await api.put<Record<string, unknown>>(`/agents/${id}`, toSnakeCase(data));
    return toCamelCase(response.data);
  },

  // Delete agent
  async delete(id: string): Promise<void> {
    await api.delete(`/agents/${id}`);
  },

  // Get default agent
  async getDefault(): Promise<Agent> {
    const response = await api.get<Record<string, unknown>>('/agents/default');
    return toCamelCase(response.data);
  },

  // Get available models
  async listModels(): Promise<string[]> {
    const response = await api.get<string[]>('/agents/models');
    return response.data;
  },

  // Get agent's effective working directory
  async getWorkingDirectory(id: string): Promise<{ path: string; isGlobalMode: boolean }> {
    const response = await api.get<{ path: string; is_global_mode: boolean }>(
      `/agents/${id}/working-directory`
    );
    return {
      path: response.data.path,
      isGlobalMode: response.data.is_global_mode,
    };
  },
};
